uisdk/src/main/java/com/karhoo/uisdk/screen/booking/booking/bookingrequest/BookingRequestView.kt

```diff
@@ -301,7 +301,7 @@ class BookingRequestView @JvmOverloads constructor(context: Context,
                 .setPositiveButton(R.string.add_card) { dialog, _ ->
                     cancelButton.isEnabled = true
                     bookingRequestButton.onLoadingComplete()
-                    showPaymentUI()
+                    handleChangeCard()
```

Comment:
Nice one. Can we add unit tests for this logic, please?

My suggestion is, to add a couple of methods to the Presenter e.g. onPaymentFailureDialogPositive() , onPaymentFailureDialogCanceled() 

Then the presenter can instruct the view on what to do (which would be exactly what you have here)

As a result, you can add a couple of unit tests to the Presenter to verify the logic will happen ... this would enable us to avoid similar bugs when we do another refactor.